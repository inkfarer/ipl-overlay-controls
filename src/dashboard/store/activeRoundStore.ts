import { ActiveRound, SwapColorsInternally } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { GameWinner } from 'types/enums/gameWinner';
import { SetActiveColorRequest, SetWinnerRequest, SwapRoundColorRequest } from 'types/messages/activeRound';
import { SetRoundRequest } from 'types/messages/rounds';
import { ActiveRoundGame } from 'types/activeRoundGame';

const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export const activeRoundReps = [ activeRound, swapColorsInternally ];

export interface ActiveRoundStore {
    activeRound: ActiveRound;
    swapColorsInternally: SwapColorsInternally;
}

export const activeRoundStore = createStore<ActiveRoundStore>({
    state: {
        activeRound: null,
        swapColorsInternally: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        setWinner(store, { winner }: { winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner });
        },
        removeWinner(): void {
            nodecg.sendMessage('removeWinner');
        },
        setActiveColor(store, request: SetActiveColorRequest): void {
            nodecg.sendMessage('setActiveColor', request);
        },
        swapColors(): void {
            swapColorsInternally.value = !swapColorsInternally.value;
        },
        setWinnerForIndex(store, { index, winner }: { index: number, winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner, roundIndex: index } as SetWinnerRequest);
        },
        resetActiveRound(): void {
            nodecg.sendMessage('resetActiveRound');
        },
        setActiveRound(store, request: SetRoundRequest): void {
            nodecg.sendMessage('setActiveRound', request);
        },
        swapRoundColor(store, request: SwapRoundColorRequest): void {
            nodecg.sendMessage('swapRoundColor', request);
        },
        updateActiveGames(store, games: ActiveRoundGame[]): void {
            nodecg.sendMessage('updateActiveGames', { games });
        },
        switchToNextColor(): void {
            nodecg.sendMessage('switchToNextColor');
        },
        switchToPreviousColor(): void {
            nodecg.sendMessage('switchToPreviousColor');
        },
        async getNextAndPreviousColors(): Promise<void> {
            return nodecg.sendMessage('getNextAndPreviousColors');
        }
    }
});

export const activeRoundStoreKey: InjectionKey<Store<ActiveRoundStore>> = Symbol();

export function useActiveRoundStore(): Store<ActiveRoundStore> {
    return useStore(activeRoundStoreKey);
}
