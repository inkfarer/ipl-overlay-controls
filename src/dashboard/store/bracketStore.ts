import { Bracket } from '@tourneyview/common';
import { defineStore } from 'pinia';

const bracketData = nodecg.Replicant<Bracket | undefined>('bracketData');

export const bracketReps = [ bracketData ];

export interface BracketStore {
    bracketData: Bracket | undefined;
}

export const useBracketStore = defineStore('brackets', {
    state: () => ({
        bracketData: undefined
    } as BracketStore),
    actions: {
        resetBracketData() {
            bracketData.value = undefined;
        }
    }
});
