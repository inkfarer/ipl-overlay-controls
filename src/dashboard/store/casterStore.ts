import { NodeCGBrowser } from 'nodecg/browser';
import { Caster, Casters, RadiaSettings } from 'schemas';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

const casters = nodecg.Replicant<Casters>('casters');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');

export const castersReps = [ casters, radiaSettings ];

export interface CasterStore {
    casters: Casters;
    uncommittedCasters: Casters;
    radiaSettings: RadiaSettings;
}

const defaultCaster = {
    name: 'New Caster',
    twitter: '',
    pronouns: '?/?'
};

export const useCasterStore = defineStore('casters', {
    state: () => ({
        casters: {},
        uncommittedCasters: {},
        radiaSettings: {
            enabled: null,
            guildID: null,
            updateOnImport: null
        }
    } as CasterStore),
    actions: {
        updateCaster({ id, newValue }: { id: string, newValue: Caster }): void {
            casters.value[id] = newValue;
        },
        removeUncommittedCaster(id: string): void {
            delete this.uncommittedCasters[id];
        },
        addUncommittedCaster({ id, caster }: { id: string, caster?: Caster }): void {
            this.uncommittedCasters[id] = caster ?? defaultCaster;
        },
        addDefaultCaster(): string {
            const id = generateId();
            this.addUncommittedCaster({ id });
            return id;
        },
        async saveUncommittedCaster({ id, caster }: { id: string, caster: Caster }): Promise<string> {
            const newId = await nodecg.sendMessage('saveCaster', caster);
            this.removeUncommittedCaster(id);
            return newId;
        },
        async loadCastersFromVc(): Promise<void> {
            const result = await nodecg.sendMessage('getLiveCommentators');
            if (result.extra && result.extra.length > 0) {
                for (let i = 0; i < result.extra.length; i++) {
                    const extraCaster = result.extra[i];
                    const id = extraCaster.discord_user_id;
                    delete extraCaster.discord_user_id;
                    this.addUncommittedCaster({ id, caster: extraCaster });
                }
            }
        },
        showCasters() {
            nodecg.sendMessage('mainShowCasters');
        },
        async removeCaster(id: string): Promise<void> {
            return nodecg.sendMessage('removeCaster', { id });
        }
    }
});
