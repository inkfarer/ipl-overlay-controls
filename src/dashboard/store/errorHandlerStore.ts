import { App } from 'vue';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    }),
    actions: {
        removeRecentError({ key }: { key: string }): void {
            delete this.recentErrors[key];
        },
        handleError({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err);

            if (Object.keys(this.recentErrors).length >= 2) return;

            const id = generateId();
            this.recentErrors[id] = err;
            window.setTimeout(() => {
                delete this.recentErrors[id];
            }, 25000);
        }
    }
});

export function setUpErrorHandler(app: App<unknown>): void {
    const store = useErrorHandlerStore();
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info });
    };
}
