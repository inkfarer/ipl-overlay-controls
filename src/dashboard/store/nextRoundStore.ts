import { NextRound } from 'schemas';
import { SetNextRoundRequest } from 'types/messages/rounds';
import { BeginNextMatchRequest } from 'types/messages/activeRound';
import { defineStore } from 'pinia';

const nextRound = nodecg.Replicant<NextRound>('nextRound');

export const nextRoundReps = [ nextRound ];

export interface NextRoundStore {
    nextRound: NextRound;
}

export const useNextRoundStore = defineStore('nextRound', {
    state: () => ({
        nextRound: null
    } as NextRoundStore),
    actions: {
        setShowOnStream(newValue: boolean) {
            nextRound.value.showOnStream = newValue;
        },
        beginNextMatch(request: BeginNextMatchRequest): void {
            nodecg.sendMessage('beginNextMatch', request);
        },
        setNextRound(request: SetNextRoundRequest): void {
            nodecg.sendMessage('setNextRound', request);
        }
    }
});
