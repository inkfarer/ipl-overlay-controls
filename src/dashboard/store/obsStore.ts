import { NodeCGBrowser } from 'nodecg/browser';
import { ObsCredentials, ObsData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetObsDataRequest } from 'types/messages/obs';

const obsData = nodecg.Replicant<ObsData>('obsData');
const obsCredentials = nodecg.Replicant<ObsCredentials>('obsCredentials');

export const obsReps = [obsData, obsCredentials];

export interface ObsStore {
    obsData: ObsData
    obsCredentials: ObsCredentials
}

export const obsStore = createStore<ObsStore>({
    state: {
        obsData: null,
        obsCredentials: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async connect(store, { address, password }: { address: string, password?: string }): Promise<void> {
            return nodecg.sendMessage('connectToObs', { address, password });
        },
        async setData(store, data: SetObsDataRequest): Promise<void> {
            return nodecg.sendMessage('setObsData', data);
        },
        async startGame(): Promise<void> {
            return nodecg.sendMessage('startGame');
        },
        async endGame(): Promise<void> {
            return nodecg.sendMessage('endGame');
        },
        setEnabled(store, enabled: boolean): void {
            nodecg.sendMessage('setObsSocketEnabled', enabled);
        }
    }
});

export const obsStoreKey: InjectionKey<Store<ObsStore>> = Symbol();

export function useObsStore(): Store<ObsStore> {
    return useStore(obsStoreKey);
}
