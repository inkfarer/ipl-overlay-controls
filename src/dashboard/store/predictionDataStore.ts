import { NodeCGBrowser } from 'nodecg/browser';
import { Prediction, PredictionStore } from 'schemas';
import { PredictionStatus } from 'types/enums/predictionStatus';
import { defineStore } from 'pinia';

const predictionStore = nodecg.Replicant<PredictionStore>('predictionStore');

export const predictionReps = [ predictionStore ];

export interface PredictionDataStore {
    predictionStore: PredictionStore;
}

export const usePredictionDataStore = defineStore('prediction', {
    state: () => ({
        predictionStore: null,
    } as PredictionDataStore),
    actions: {
        async lockPrediction() {
            return patchPrediction(
                this.predictionStore.currentPrediction,
                PredictionStatus.LOCKED,
                [ PredictionStatus.ACTIVE ]);
        },
        async cancelPrediction() {
            return patchPrediction(
                this.predictionStore.currentPrediction,
                PredictionStatus.CANCELED,
                [ PredictionStatus.ACTIVE, PredictionStatus.LOCKED ]);
        },
        async resolvePrediction({ winningOutcomeIndex }: { winningOutcomeIndex: number }) {
            if (winningOutcomeIndex > 1 || winningOutcomeIndex < 0) {
                throw new Error(`Cannot resolve prediction with outcome index ${winningOutcomeIndex}.`);
            } else if (!this.predictionStore.currentPrediction?.id) {
                throw new Error('No prediction to resolve.');
            } else if (this.predictionStore.currentPrediction.status !== PredictionStatus.LOCKED ) {
                throw new Error('Can only resolve a locked prediction.');
            }

            return nodecg.sendMessage('patchPrediction', {
                id: this.predictionStore.currentPrediction.id,
                status: PredictionStatus.RESOLVED,
                winning_outcome_id: this.predictionStore.currentPrediction.outcomes[winningOutcomeIndex].id
            });
        },
        async createPrediction(
            { title, teamAName, teamBName, duration }:
                { title: string, teamAName: string, teamBName: string, duration: number }
        ) {
            if (
                [PredictionStatus.ACTIVE, PredictionStatus.LOCKED]
                    .includes(this.predictionStore.currentPrediction?.status as PredictionStatus)
            ) {
                throw new Error('An unresolved prediction already exists.');
            }

            return nodecg.sendMessage('postPrediction', {
                title,
                outcomes: [
                    { title: teamAName },
                    { title: teamBName }
                ],
                prediction_window: duration
            });
        },
        async reconnect() {
            return nodecg.sendMessage('reconnectToRadiaSocket');
        }
    }
});

async function patchPrediction(
    currentPrediction: Prediction,
    status: PredictionStatus,
    allowedForStatuses: PredictionStatus[]
): Promise<void> {
    if (!currentPrediction || !allowedForStatuses.includes(currentPrediction?.status as PredictionStatus)) {
        throw new Error(`Cannot set prediction status at this time. Status must match one of [${allowedForStatuses.join(', ')}]`);
    }

    return nodecg.sendMessage('patchPrediction', {
        id: currentPrediction.id,
        status: status,
    });
}
