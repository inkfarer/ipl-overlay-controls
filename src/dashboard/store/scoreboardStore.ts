import { NodeCGBrowser } from 'nodecg/browser';
import { ScoreboardData } from 'schemas';
import { defineStore } from 'pinia';

const scoreboardData = nodecg.Replicant<ScoreboardData>('scoreboardData');

export const scoreboardReps = [scoreboardData];

export interface ScoreboardStore {
    scoreboardData: ScoreboardData
}

export const useScoreboardStore = defineStore('scoreboard', {
    state: () => ({
        scoreboardData: {
            flavorText: '',
            isVisible: null
        }
    } as ScoreboardStore),
    actions: {
        setFlavorText(flavorText: string): void {
            scoreboardData.value.flavorText = flavorText;
        },
        setScoreboardVisible(isVisible: boolean): void {
            scoreboardData.value.isVisible = isVisible;
        }
    }
});
