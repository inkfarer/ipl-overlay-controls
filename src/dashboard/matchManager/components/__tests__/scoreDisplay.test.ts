import ScoreDisplay from '../scoreDisplay.vue';
import { useActiveRoundStore } from '../../../store/activeRoundStore';
import { config, mount } from '@vue/test-utils';
import { GameWinner } from 'types/enums/gameWinner';
import { createTestingPinia, TestingPinia } from '@pinia/testing';
import { IplButton } from '@iplsplatoon/vue-components';

describe('ScoreDisplay', () => {
    let pinia: TestingPinia;

    beforeEach(() => {
        pinia = createTestingPinia();

        useActiveRoundStore().$state = {
            activeRound: {
                teamA: {
                    score: 0,
                    id: null,
                    name: null,
                    showLogo: null,
                    players: null,
                    color: null
                },
                teamB: {
                    score: 2,
                    id: null,
                    name: null,
                    showLogo: null,
                    players: null,
                    color: null
                },
                activeColor: null,
                match: null,
                games: [
                    {
                        winner: GameWinner.BRAVO,
                        stage: null,
                        mode: null
                    },
                    {
                        winner: GameWinner.BRAVO,
                        stage: null,
                        mode: null
                    },
                    {
                        winner: GameWinner.NO_WINNER,
                        stage: null,
                        mode: null
                    },
                ],
            },
            swapColorsInternally: false
        };
    });

    config.global.stubs = {
        IplButton: true
    };

    it('matches snapshot', () => {
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('disables adding score if last game in round has been completed', () => {
        const store = useActiveRoundStore();
        store.activeRound.teamA.score = 1;
        store.activeRound.teamB.score = 2;
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        expect(wrapper.getComponent('[data-test="team-a-plus-btn"]').attributes().disabled).toEqual('true');
        expect(wrapper.getComponent('[data-test="team-b-plus-btn"]').attributes().disabled).toEqual('true');
    });

    it('disables removing team A score if last game was won by team B', () => {
        const store = useActiveRoundStore();
        store.activeRound.games = [
            {
                winner: GameWinner.BRAVO,
                stage: null,
                mode: null
            },
            {
                winner: GameWinner.NO_WINNER,
                stage: null,
                mode: null
            },
            {
                winner: GameWinner.NO_WINNER,
                stage: null,
                mode: null
            },
        ];
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        expect(wrapper.getComponent('[data-test="team-a-minus-btn"]').attributes().disabled).toEqual('true');
        expect(wrapper.getComponent('[data-test="team-b-minus-btn"]').attributes().disabled).toEqual('false');
    });

    it('disables removing team B score if last game was won by team A', () => {
        const store = useActiveRoundStore();
        store.activeRound.games = [
            {
                winner: GameWinner.BRAVO,
                stage: null,
                mode: null
            },
            {
                winner: GameWinner.ALPHA,
                stage: null,
                mode: null
            },
            {
                winner: GameWinner.NO_WINNER,
                stage: null,
                mode: null
            },
        ];
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        expect(wrapper.getComponent('[data-test="team-a-minus-btn"]').attributes().disabled).toEqual('false');
        expect(wrapper.getComponent('[data-test="team-b-minus-btn"]').attributes().disabled).toEqual('true');
    });

    it('disables removing score if round has no completed games', () => {
        const store = useActiveRoundStore();
        store.activeRound.games = [
            {
                winner: GameWinner.NO_WINNER,
                stage: null,
                mode: null
            },
            {
                winner: GameWinner.NO_WINNER,
                stage: null,
                mode: null
            },
            {
                winner: GameWinner.NO_WINNER,
                stage: null,
                mode: null
            },
        ];
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        expect(wrapper.getComponent('[data-test="team-a-minus-btn"]').attributes().disabled).toEqual('true');
        expect(wrapper.getComponent('[data-test="team-b-minus-btn"]').attributes().disabled).toEqual('true');
    });

    it('sends message to store when adding points to team A', () => {
        const store = useActiveRoundStore();
        store.setWinner = jest.fn();
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        wrapper.getComponent<typeof IplButton>('[data-test="team-a-plus-btn"]').vm.$emit('click');

        expect(store.setWinner).toHaveBeenCalledWith({ winner: GameWinner.ALPHA });
    });

    it('sends message to store when adding points to team B', () => {
        const store = useActiveRoundStore();
        store.setWinner = jest.fn();
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        wrapper.getComponent<typeof IplButton>('[data-test="team-b-plus-btn"]').vm.$emit('click');

        expect(store.setWinner).toHaveBeenCalledWith({ winner: GameWinner.BRAVO });
    });

    it('sends message to store when removing points from team A', () => {
        const store = useActiveRoundStore();
        store.removeWinner = jest.fn();
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        wrapper.getComponent<typeof IplButton>('[data-test="team-a-minus-btn"]').vm.$emit('click');

        expect(store.removeWinner).toHaveBeenCalled();
    });

    it('sends message to store when removing points from team B', () => {
        const store = useActiveRoundStore();
        store.removeWinner = jest.fn();
        const wrapper = mount(ScoreDisplay, {
            global: {
                plugins: [pinia]
            }
        });

        wrapper.getComponent<typeof IplButton>('[data-test="team-b-minus-btn"]').vm.$emit('click');

        expect(store.removeWinner).toHaveBeenCalled();
    });
});
