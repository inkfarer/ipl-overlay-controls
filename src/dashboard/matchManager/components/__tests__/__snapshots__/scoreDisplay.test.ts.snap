// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreDisplay matches snapshot 1`] = `
"<div data-v-11cc5f1c="" class="ipl-space color-dark layout horizontal center-horizontal center-vertical">
  <div data-v-11cc5f1c="" class="ipl-space color-light layout horizontal score-display-space">
    <div class="layout vertical">
      <ipl-button-stub icon="plus" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="true" disableonsuccess="false" href="" data-test="team-a-plus-btn"></ipl-button-stub>
      <ipl-button-stub icon="minus" color="red" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="true" disableonsuccess="false" href="" class="m-t-4" data-test="team-a-minus-btn"></ipl-button-stub>
    </div>
    <div class="layout horizontal center-horizontal center-vertical score-wrapper left"><span class="score">0</span></div>
  </div><span class="score-separator">:</span>
  <div data-v-11cc5f1c="" class="ipl-space color-light layout horizontal score-display-space">
    <div class="layout horizontal center-horizontal center-vertical score-wrapper right"><span class="score">2</span></div>
    <div class="layout vertical">
      <ipl-button-stub icon="plus" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="true" disableonsuccess="false" href="" data-test="team-b-plus-btn"></ipl-button-stub>
      <ipl-button-stub icon="minus" color="red" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="true" disableonsuccess="false" href="" class="m-t-4" data-test="team-b-minus-btn"></ipl-button-stub>
    </div>
  </div>
</div>"
`;
